lib/JunkDetection/CXX/ASTStorage.cpp

```diff
@@ -136,3 +136,47 @@ ThreadSafeASTUnit *ASTStorage::findAST(const MutationPoint *point) {
   astUnits[sourceFile] = std::unique_ptr<ThreadSafeASTUnit>(threadSafeAST);
   return threadSafeAST;
 }
+
+MutationPointSourceInfo
+ASTStorage::getSourceInfo(MutationPoint *mutationPoint) {
+  MutationPointSourceInfo info;
+  clang::SourceRange sourceRange;
+
+  if (sourceRanges.count(mutationPoint) == 0) {
+    // TODO: I would prefer to have an optional here rather than returning
+    // empty struct.
+    return info;
+  }
+
+  sourceRange = sourceRanges[mutationPoint];
+
+  ThreadSafeASTUnit *astUnit = findAST(mutationPoint);
+  clang::ASTContext &astContext = astUnit->getASTContext();
+  clang::SourceManager &sourceManager = astUnit->getSourceManager();
+
+  clang::SourceLocation sourceLocationBegin = sourceRange.getBegin();
+  clang::SourceLocation sourceLocationEnd = sourceRange.getEnd();
+
+  /// Clang AST: how to get more precise debug information in certain cases?
+  /// http://clang-developers.42468.n3.nabble.com/Clang-AST-how-to-get-more-precise-debug-information-in-certain-cases-td4065195.html
+  /// https://stackoverflow.com/questions/11083066/getting-the-source-behind-clangs-ast
+  clang::SourceLocation sourceLocationEndActual =
+      clang::Lexer::getLocForEndOfToken(sourceLocationEnd, 0, sourceManager,
+                                        astContext.getLangOpts());
+
+  info.beginColumn =
+      sourceManager.getExpansionColumnNumber(sourceLocationBegin);
+  info.beginLine =
+      sourceManager.getExpansionLineNumber(sourceLocationBegin, nullptr);
+  info.endColumn =
+      sourceManager.getExpansionColumnNumber(sourceLocationEndActual);
+  info.endLine =
+      sourceManager.getExpansionLineNumber(sourceLocationEndActual, nullptr);
+
+  return info;
+}
```

Comment:
As mentioned earlier, it does not belong here. Unless there is a strong reason that I'm missing.

Reply:
This will change, let's agree on setting up the AST nodes first.